nip-nostr-connect.md

```diff
@@ -0,0 +1,130 @@
+NIP-Nostr-Connect
+======
+
+Nostr Connect allows Nostr apps to connect with remote signing devices
+------------------------
+
+`draft` `optional` `author:tiero` `author:giowe` `author:vforvalerio87`
+
+## Rationale
+
+Having to enter your Nostr private key on each website or random app _sucks_.
+
+
+## `TL;DR`
+
+
+**App** (*typically a web app, it generates a random ephemeral keypair*) and **Wallet** (*typically a mobile app, it holds the private key of the user that represents his Nostr account*) send to each other `kind:4` encrypted DMs, using a relay of choice. 
+
+App prompts the Wallet to do things such as fetching the public key or signing events.
+
+The `content` field must be a JSONRPC-ish **request** or **response**.
+
+## Protocol
+
+### Messages
+
+#### Request
+
+```json
+{
+  "id": <random_string>,
+  "method": <one_of_the_methods>,
+  "params": [<anything>, <else>]
+}
+```
+
+#### Response
+
+```json
+{
+  "id": <request_id>,
+  "result": <anything>,
+  "error": <reason>
+}
+```
+
+### Methods
+
+- `connect`
+  - params [`pubkey`]
+- `disconnect`
+  - params []
+- `get_public_key`
+  - params []
+  - result `pubkey`
+- `sign_event`
+  - params [`event`]
+  - result `signature`
+
+#### optional
+
+- `delegate`
+- `get_relays`
+- `nip04_encrypt`
+- `nip04_decrypt`
+
+
+### Nostr Connect URI
+
+**Wallet** discovers **App** by scanning a QR code or clicking on a deep link or copy-pasting an URI.
+
+The **App** generates a special URI with prefix `nostr://` and base path `connect` with the following querystring parameters
+
+- `target` hexadecimal public key of the **App**
+- `relay` URL of the relay of choice where the **App** is connected and the **Wallet** must send and listen for messages.
+- `metadata`  metadata JSON of the **App** 
+    - `url` URL of the website requesting the connection
+    - `name` human-readable name of the **App** 
+    - `description` (optional) description of the **App**
+    - `icons` (optional) array of URLs for icons of the **App**.
+
+#### Example
+
+```sh
+nostr://connect?target=<pubkey>&relay=<relay>&metadata={"url": "example.com","name": "Example"}
```

Comment:
Good catch, thanks! totally forgot about it!

Reply:
I still don't like this metadata bit. Maybe let's use just a "name" so it is clearer that it is not a checked thing? A URL can still be phished and some people will find that more trustworthy and be misled.

Reply:
Now I am thinking: doesn't the app requesting this need a pubkey of its own? The way to get metadata from that would be through Nostr then, no? The signer could request the `kind: 0` event for that pubkey from the same relay and possibly even validate a NIP-05 to show a domain name.

Reply:
I like this, sounds clean

Reply:
But the pubkey of the app is basically generated by the client (in the user browser) and could potentially change accross sessions.

I am failing to see how publishing an event before showing the connect URI, would prevent any phishing? Or Im  loosing some piece here?

Reply:
> Maybe let's use just a "name"

yes I think removing the URL it's best we can do here.

> I still don't like this metadata bit. 

I think we can make this process a bit more interactive, so to force user to "double-check"

The Signer instead of asking the user "<metadata.name> wants to connect..." we can:

- Generate QRCode and show in the App UI a sequence of emoji derived from the App's pubkey
- In the Signer UI, it shows a radio button with three choices of emoji sequence
- The user must pick the same sequence he sees on the App screen (two are random, one is derived by the pubkey in the Connect URI)

This has the job of "forcing" a bit the user to take the time to double-check the website he's connecting to

PS: we can do number instead of emoji, if we want to be more like boomers and less genZ

Reply:
I guess this is not something the protocol should be aware of anyway, it's something the Signer app does on top anyway since the emoji/number sequence to pick is derived from the pubkey encoded in the Connect URI.

I would eventually implement it in the SDK, but not specify it here in the nip.

Is this acceptable to you guys?